api/turing/service/router_deployment_service.go

```diff
@@ -251,6 +251,15 @@ func (ds *deploymentService) UndeployRouterVersion(
 		return err
 	}
 
+	// Create a new docker config for the ensembler with the newly generated image
+	if routerVersion.Ensembler != nil && routerVersion.Ensembler.Type == models.EnsemblerPyFuncType {
+		// An empty image ref is passed since it is not used by the Kubernetes controller for undeployment
+		routerVersion.Ensembler.DockerConfig = generateDockerConfigFromPyfuncConfig(
```

Comment:
I'd have expected the `routerVersion.Ensembler.DockerConfig` to be set at this point already, as we are passing the routerVersion reference to the [DeployRouterVersion](https://github.com/caraml-dev/turing/blob/main/api/turing/service/router_deployment_service.go#L110) method (which will alter the docker config after the image is built) and the `UndeployRouterVersion` is called downstream using the same routerVersion ref: https://github.com/caraml-dev/turing/blob/main/api/turing/api/deployment_controller.go#L76

What am I missing?

(Your approach to fix this bug will work too, although it's making the assumption that there will be a docker config attached to the Pyfunc config always, even in cases where the image building may have timed out. Nothing wrong with this but  wondering why we are unable to propagate the docker config down from the deploy step.)

Reply:
Ah that's true if the `UndeployRouterVersion` method is only called when updating routers (i.e. deploying a new router version and then deleting the old one) via `deployOrRollbackRouter`, but the  `UndeployRouterVersion` method is also used to undeploy routers using the undeploy endpoint via `undeployRouter` [here](https://github.com/caraml-dev/turing/blob/main/api/turing/api/deployment_controller.go#L266). In the second case, the `routerVersion.Ensembler.DockerConfig`(-s) wouldn't have been set since they would've been [retrieved](https://github.com/caraml-dev/turing/blob/main/api/turing/api/deployment_controller.go#L292) from the DB. 

Although that said, it's possible to introduce more logic into the conditional statement above to only re-generate the docker config if it's missing (i.e. when we are only performing an undeployment) and instead use the existing docker config if it exists in the `routerVersion` (i.e. when we are performing a router update).

> although it's making the assumption that there will be a docker config attached to the Pyfunc config always, even in cases where the image building may have timed out

It's not a pretty assumption but I guess it's partly linked to 3 things, 1) that we are rebuilding entire `KnativeService` structs for undeploying Knative services, which is a little redundant since all we need is the Knative service name, 2) that we aren't storing any information in the `routerVersion` struct/data/schema about the build success/failure of its pyfunc ensembler and 3) that we an 'ensembler' Knative service and the methods that correspond to it are ultimately tightly coupled to the presence of a docker config and it isn't generic enough to operate on a pyfunc config. 

So in order to re-use the existing workflow for undeploying the pyfunc ensemblers, we'll need to generate a docker config temporarily, so that we can build an ensembler Knative service out of it, so that the controller's able to undeploy it using the Knative name contained within the `KnativeService` struct. That's like the approach I thought of but maybe you were seeing another approach that I may have missed out 😅

Reply:
> but the UndeployRouterVersion method is also used to undeploy routers using the undeploy endpoint via undeployRouter [here](https://github.com/caraml-dev/turing/blob/main/api/turing/api/deployment_controller.go#L266)

In this scenario, the docker_config should have been saved to the DB (we also display it in the UI for pyfunc ensemblers). 🤔 

There are 3 undeploy scenarios in all:
* Call `/undeploy` endpoint
* Update a router version successfully (then the existing router version is undeployed)
* Update a router version but it's unsuccessful (then, the new router version will be undeployed immediately)

In scenarios 1 and 2, we should have the docker_config in the DB and in scenario 3, it should be in the routerVersion object ref that's passed around. 😕

Reply:
Okay I think we've found the true bug behind this to be Gorm, so the ultimate fix to this is to update the Gorm configs to allow all nested fields to be updated whenever we perform a `Save` operation on the DB. I've updated the PR description with the new details; thanks for helping to dig into this! :D

Reply:
Thanks for the fix, @deadlycoconuts (and also for updating all the details in the PR description). LGTM!